src/architectures/encoder.rs

```diff
@@ -13,26 +14,77 @@ pub struct EncoderOutput {
 }
 
 impl EncoderOutput {
+    /// Create an encoder output.
     pub fn new(all_outputs: Vec<Tensor>) -> Self {
         Self { all_outputs }
     }
 }
 
 impl LayerOutputs for EncoderOutput {
+    /// All layer outputs.
     fn layer_outputs(&self) -> &[Tensor] {
         &self.all_outputs
     }
 
+    /// Embedding layer output.
     fn embedding_layer_output(&self) -> Option<&Tensor> {
         self.all_outputs.first()
     }
 }
 
+/// Trait for encoders.
+pub trait Encoder {
+    /// Encode an input sequence.
+    ///
+    /// Returns the encoder output.
+    ///
+    /// * `piece_ids` - Input sequence.
+    ///   *Shape:* `(batch_size, seq_len)`
+    /// * `attention_mask` - Attention mask. Sequence elements for which the
+    ///   corresponding mask element is set to `false` are ignored during
+    ///   attention calculation.
+    ///   *Shape:* `(batch_size, seq_len)`
+    /// * `positions` - Input positions.
+    ///   *Shape:* `(batch_size, seq_len)`
+    /// * `type_ids` - Input type ids.
+    ///   *Shape:* `(batch_size, seq_len)`
+    /// * `train` - Whether to train the layer.
+    fn forward_t(
+        &self,
+        piece_ids: &Tensor,
+        attention_mask: &AttentionMask,
+        positions: Option<&Tensor>,
+        type_ids: Option<&Tensor>,
+        train: bool,
```

Comment:
`train` [should come before](https://github.com/oxidized-transformers/oxidized-transformers/blob/main/CODING.md#module-forward-signature) the optional arguments, I think.

Reply:
Yeah, as you suggest, it's probably better for me to do this in a separate PR, so I can go through my dev branch first and all the existing models on the branch get updated through the refactor.